flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java

```diff
@@ -104,9 +103,19 @@ private Table convertConnectorTable(ConnectorCatalogTable<?, ?> table) {
 	}
 
 	private Table convertCatalogTable(ObjectPath tablePath, CatalogTable table) {
+		TableSource<?> tableSource;
 		Optional<TableFactory> tableFactory = catalog.getTableFactory();
-		TableSource<Row> tableSource = tableFactory.map(tf -> ((TableSourceFactory) tf).createTableSource(tablePath, table))
-			.orElse(TableFactoryUtil.findAndCreateTableSource(table));
+		if (tableFactory.isPresent()) {
+			TableFactory tf = tableFactory.get();
+			if (tf instanceof TableSourceFactory) {
+				tableSource = ((TableSourceFactory) tf).createTableSource(tablePath, table);
+			} else {
+				throw new TableException(String.format("TableFactory provided by catalog %s must implement TableSourceFactory",
```

Comment:
minor: shall we use similar error message as that in `convertConnectorTable()`, something like "Cannot query a sink-only table. TableFactory provided by catalog %s must implement ....".

Asking this because when I first saw this exception, I got a bit on confused why it cannot be `TableSinkFactory`.